internal/core/model/adapters.rs

```diff
@@ -413,3 +415,510 @@ fn test_filter_model() {
+/// assert_eq!(sorted_model.row_data(0).unwrap(), 1);
+/// assert_eq!(sorted_model.row_data(1).unwrap(), 3);
+/// assert_eq!(sorted_model.row_data(2).unwrap(), 5);
+/// ```
+///
+/// Alternatively you can use the shortcut [`ModelExt::sort`].
+/// ```
+/// # use slint::{Model, ModelExt, VecModel, SharedString, SortModel};
+/// let sorted_model = VecModel::from(vec![
+///     5,
+///     1,
+///     3,
+/// ]).sort();
+/// # assert_eq!(sorted_model.row_data(0).unwrap(), 1);
+/// # assert_eq!(sorted_model.row_data(1).unwrap(), 3);
+/// # assert_eq!(sorted_model.row_data(2).unwrap(), 5);
+/// ```
+///
+/// If you want to modify the underlying [`VecModel`] you can give it a [`Rc`] of the SortModel:
+/// ```
+/// # use std::rc::Rc;
+/// # use slint::{Model, VecModel, SharedString, SortModel};
+/// let model = Rc::new(VecModel::from(vec![
+///     SharedString::from("Lorem"),
+///     SharedString::from("ipsum"),
+///     SharedString::from("dolor"),
+/// ]));
+///
+/// let sorted_model = SortModel::new(model.clone(), |lhs, rhs| lhs.to_lowercase().cmp(&rhs.to_lowercase()));
+///
+/// assert_eq!(sorted_model.row_data(0).unwrap(), SharedString::from("dolor"));
+/// assert_eq!(sorted_model.row_data(1).unwrap(), SharedString::from("ipsum"));
+/// assert_eq!(sorted_model.row_data(2).unwrap(), SharedString::from("Lorem"));
+///
+/// model.set_row_data(1, SharedString::from("opsom"));
+///
+/// assert_eq!(sorted_model.row_data(0).unwrap(), SharedString::from("dolor"));
+/// assert_eq!(sorted_model.row_data(1).unwrap(), SharedString::from("Lorem"));
+/// assert_eq!(sorted_model.row_data(2).unwrap(), SharedString::from("opsom"));
+/// ```
+pub struct SortModel<M>(Pin<Box<ModelChangeListenerContainer<SortModelInner<M>>>>)
+where
+    M: Model + 'static;
+
+impl<M> SortModel<M>
+where
+    M: Model + 'static,
+{
+    /// Creates a new SortModel based on the given `wrapped_model` and sorted by `sort_function`.
+    /// Alternativly you can use [`ModelExt::sort_by`] on your Model.
+    pub fn new<F>(wrapped_model: M, sort_function: F) -> Self
+    where
+        F: FnMut(&M::Data, &M::Data) -> std::cmp::Ordering + 'static,
+    {
+        let sorted_model_inner = SortModelInner {
+            wrapped_model,
+            sort_helper: Box::new(FnSortHelper {
+                sort_function: RefCell::new(sort_function),
+                _type: PhantomData::default(),
+            }),
+            mapping: RefCell::new(Vec::new()),
+            notify: Default::default(),
+        };
+
+        sorted_model_inner.build_mapping_vec();
+
+        let container = Box::pin(ModelChangeListenerContainer::new(sorted_model_inner));
+
+        container.wrapped_model.model_tracker().attach_peer(container.as_ref().model_peer());
+
+        Self(container)
+    }
+
+    /// Creates a new SortModel based on the given `wrapped_model` and sorted in ascending order.
+    /// Alternativly you can use [`ModelExt::sort`] on your Model.
+    pub fn new_ascending(wrapped_model: M) -> Self
```

Comment:
I would call this `new` and the one above maybe `new_with_custom_sort` or something.